### ISSUE TITLE
Bundle.request bug relating to PR#74

### ISSUE THREAD

**Issue by bkaney** [microsoft/FHIR-Converter#76]
**Describe the bug**
There is a problem in https://github.com/microsoft/FHIR-Converter/pull/74

This should have remained a `POST` but remove the `{{ID}}` part from the url (just the resource type). 

A common pattern in FHIR Bundle transactions is to let the server mint the new logic id's as the transaction is committed, while maintaining referential integrity. See https://www.hl7.org/fhir/http.html#trules

**Comment by ranvijaykumar** [comment]
While using POST, the server ignores the provided resource id and generates a new one. For example, if we POST the same patient resource multiple times, every time the server will create a resource with new resource id. This will result in duplicate records in the server for the same physical entity. This is a very real scenario. Imagine for the same patient we receive 5 HL7 messages, each containing a PID segment. During conversion process, we will create 5 patient resources. If we POST all the 5 resources, the patient will 5 different resource ids in the FHIR server. 

On the other hand, PUT preserves the Id provided in the resource and behaves like UPSERT. In the above scenario, we will get only one patient resource in the FHIR server.

**Comment by bkaney** [comment]
You are correct and the server generates new IDs. However within a `Bundle.transaction`, the server maintains referential integrity. This is important, because the server should be the authority in generating the IDs (and not an external service).

If you have an external service that generates IDs there could be collisions, especially as you might be importing from multiple systems. Also, the server may have it's own ID scheme (auto-incrementing, uuids, something else).

I think a better practice is to put everything in a single transaction (where you maintain transactional integrity, e.g. use the `fullUrl` for references) and then let the server assign logical ids when it commits the transaction.

For example, you can POST this to any FHIR test server (HAPI, Vonk, etc) that supports transactions (NOTE how the `Encounter.subject` uses the `fullUrl`):

```json
{
  "resourceType": "Bundle",
  "type": "transaction",
  "entry": [
    {
      "fullUrl": "urn:uuid:64bac34e-e611-3549-848b-89416176aa0b",
      "resource": {
        "resourceType": "Patient",
        "id": "64bac34e-e611-3549-848b-89416176aa0b",
        "name": [
          {
            "family": "DUCK",
            "given": [ "DONALD" ]
          }
        ],
        "birthDate": "1924-10-10",
        "gender": "male"
      },
      "request": {
        "method": "POST",
        "url": "Patient"
      }
    },
    {
      "fullUrl": "urn:uuid:4cfe8d6d-3fc8-3e41-b921-f204be18db31",
      "resource": {
        "resourceType": "Encounter",
        "id": "4cfe8d6d-3fc8-3e41-b921-f204be18db31",
        "status": "unknown",
        "subject": {
          "reference": "urn:uuid:64bac34e-e611-3549-848b-89416176aa0b"
        }
      },
      "request": {
        "method": "POST",
        "url": "Encounter"
      }
    }
  ]
}
```
```

**Comment by ranvijaykumar** [comment]
External systems can send data (HL7/CDA/..) for the same patient in different batches on different days. The data need to be processed as they come. As a result, data for the same patient will very likely go to different bundles. In order to keep the resource id consistent across bundles, fullURL could be used. However, this would work only if the server maintains a mapping between the fullURL and logical id for all the resources. Maintaining the mapping is optional, and not all servers do it. Therefore by generating PUT by default, we are giving control to users to maintain the resource ids. If it does not work for certain scenario, users can change the templates as needed.

**Comment by bkaney** [comment]
@ranvijaykumar Thanks for your reply. I understand the developer convenience in externally prescribing the logical IDs of resources, it does make life simpler. However, to prioritize patient safety, we should let the FHIR server deal with assigning id's and references. In the templates we could use `.identifier` elements to keep track of the foreign id's. Then either at at query-time, use identifier.

Then have a two different (sets of) templates for either:
1. Use conditional-create in the Bundle.transaction using the identifier as mentioned above
2. Use PUT (like you propose) if-and-only-if you know you can trust that the foreign system won't have id conflicts.

IDK. Assuming that no conflicts as the default behavior seems less safe?

**Comment by bkaney** [comment]
Pinging this thread. Another, more "standards-based" option is to use conditional-create in the bundle transaction. This is the FHIR way to do upsert while maintaining ref. integrity. The template could make use if `identifier` or maybe `meta.source` for this existence conditional logic.

See: https://www.hl7.org/fhir/http.html#cond-update and https://www.hl7.org/fhir/http.html#ccreate

**Comment by bkaney** [comment]
@ranvijaykumar here is a working example of two bundles that simulate data from two distinct messages (bundle-1 might happen today and bundle-2 tomorrow, for example). This pattern uses a combination of conditional create and conditional references is how to ensure the FHIR server maintains ref integrity. 

https://gist.github.com/bkaney/a7d197ede36d057a81d3d4e1834aee60

**Comment by ranvijaykumar** [comment]
Thanks for all your inputs. We will discuss these with the team. Prima facie, here is one challenge I see with conditional PUT. Imagine there are two resources in the FHIR server, both having the same identifier, http://hl7.org/fhir/sid/us-ssn|123121234. The possibility of having two resources with the same identifier is very real when multiple sources are sending data. Even within one system there are chances of having duplicate records. In such a case, if we use http://hl7.org/fhir/sid/us-ssn|123121234 as the identifier, the server would return 412. Then, either we need to resolve the duplication, or build a logic to pick one of the resources to update and use PUT using the resource id.

**Comment by bkaney** [comment]
You could increase the complexity in your conditional reference to try to avoid the chance of conflicts. But in any case, this is a challenging problem for sure, and I thing will vary depending on specific use-cases.

What do you think about other repo(s) separate from the tooling that could organize different community template strategies that could be forked and then easily later loaded by the tool?

**Comment by qiwjin** [comment]
closed this issue since there are no more discussion, feel to open if necessary.
